### PR TITLE
add turnpike content guard (HMS-4783)

### PR DESCRIPTION
**Pull request by loadtheaccumulator**
# Description
Add turnpike content guard for ostree and associated feature flag
Add mTLS URL swap to support turnpike switch
Switch RBAC to configurable option

FIXES: HMS-4783

## Type of change

What is it?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation update
- [ ] Tests update
- [ ] Refactor

### CODE REVIEW

**Comment by loadtheaccumulator** [comment]
I set this to draft while additional tests run, but wanted to get this up here for perusal while I'm out through Friday.

File: pkg/services/repostore/pulpstore.go [hunk 1]
```diff
@@ -168,17 +169,17 @@ func distributionURL(ctx context.Context, distBaseURL string, domain string, rep
...
+		stagePulpURL := fmt.Sprintf("%s/api/pulp-content/%s/%s", cfg.Pulp.URL, domain, repoName)
 		stageDistURL, err := url.Parse(stagePulpURL)
 		if err != nil {
 			return "", errors.New("Unable to set user:password for Pulp distribution URL")
 		}
-		stageDistURL.User = url.UserPassword(cfg.PulpContentUsername, cfg.PulpContentPassword)
+		stageDistURL.User = url.UserPassword(cfg.Pulp.ContentUsername, cfg.Pulp.ContentPassword)
 
 		distURL = stageDistURL.String()
 	}
```
Comment (lzap) [inline pkg/services/repostore/pulpstore.go:185]: This looks like a refactoring that should be in a separate commit. I do not insist, tho, having the pulp config changes separately would be perfect.

**Review by lzap — COMMENTED** [review]
This is not complete, looks great, tho I suggest to avoid new feature flags if we can.

File: pkg/clients/pulp/guards_composite.go [hunk 2]
```diff
@@ -107,12 +108,27 @@ func (ps *PulpService) ContentGuardEnsure(ctx context.Context, orgID string) (*C
...
+		}
+
+		contentGuardHref = rcg.PulpHref
 	}
 
-	ccg, err := ps.CompositeGuardEnsure(ctx, orgID, *hcg.PulpHref, *rcg.PulpHref)
+	ccg, err := ps.CompositeGuardEnsure(ctx, orgID, *hcg.PulpHref, *contentGuardHref)
 	if err != nil {
 		return nil, err
 	}
```
[the remark this answers is not in the document]
Comment (loadtheaccumulator) [reply pkg/clients/pulp/guards_composite.go:134]: Absolutely, I was thinking that last week when we talked, but took the shortcut with the flag.

File: pkg/clients/pulp/guards_turnpike.go [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+		return nil, fmt.Errorf("unexpected response: %d, body: %s", resp.StatusCode(), string(resp.Body))
+	}
+
+	return resp.JSON201, nil
+}
+
+// TurnpikeGuardEnsure ensures that the Turnpike guard is created and returns it. The method is idempotent.
+func (ps *PulpService) TurnpikeGuardEnsure(ctx context.Context) (*HeaderContentGuardResponse, error) {
+	cg, err := ps.TurnpikeGuardFind(ctx)
+	// nolint: gocritic
```
[the remark this answers is not in the document]
Comment (loadtheaccumulator) [reply pkg/clients/pulp/guards_turnpike.go:110]: I actually left that nolint in from the RBAC code. I'll run one without it.

File: pkg/services/repostore/pulpstore.go (hunk 1 above)
Comment (loadtheaccumulator) [reply pkg/services/repostore/pulpstore.go:185]: Either way works for me. Both are available. I can strip it out on next pass.

File: pkg/clients/pulp/guards_composite.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (loadtheaccumulator) [reply pkg/clients/pulp/guards_composite.go:134]: Although, it does give a clear indication which one is working this early on.

File: pkg/clients/pulp/guards_turnpike.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+// TurnpikeGuardCreate creates a new Turnpike guard and returns it.
+func (ps *PulpService) TurnpikeGuardCreate(ctx context.Context) (*HeaderContentGuardResponse, error) {
+	cfg := config.Get()
+
+	req := HeaderContentGuard{
+		Name:        TurnpikeGuardName,
+		Description: ptr.To("EDGE"),
+		HeaderName:  "x-rh-identity",
+		HeaderValue: cfg.Pulp.GuardSubjectDN,
```
Comment (lzap) [inline pkg/clients/pulp/guards_turnpike.go:90]: Can you parametrize the value like we do in `HeaderGuardCreate`?

File: pkg/clients/pulp/guards_composite.go [hunk 5]
```diff
@@ -107,12 +108,27 @@ func (ps *PulpService) ContentGuardEnsure(ctx context.Context, orgID string) (*C
...
+		rcg, err := ps.RbacGuardEnsure(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		contentGuardHref = rcg.PulpHref
 	}
 
-	ccg, err := ps.CompositeGuardEnsure(ctx, orgID, *hcg.PulpHref, *rcg.PulpHref)
+	ccg, err := ps.CompositeGuardEnsure(ctx, orgID, *hcg.PulpHref, *contentGuardHref)
```
Comment (lzap) [inline pkg/clients/pulp/guards_composite.go:131]: I am making this a variadic function so you can pass any number of hrefs here.

https://github.com/RedHatInsights/edge-api/pull/2732

File: pkg/clients/pulp/guards_composite.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (lzap) [reply pkg/clients/pulp/guards_composite.go:134]: I want us to eliminate most of the feature flags. Dropped a comment how to do this.

**Comment by lzap** [comment]
I am almost done adding MTLS, this is the last PR: https://github.com/osbuild/osbuild-composer/pull/4412

Once this is merged and promoted to stage, IB should be automatically able to reach out to any pulp repository on stage or prod as long as the DN is set accordingly in edge.

File: pkg/clients/pulp/guards_composite.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (loadtheaccumulator) [reply pkg/clients/pulp/guards_composite.go:134]: Is the intention to keep both rbac and turnpike long-term?

File: pkg/clients/pulp/guards_turnpike.go (hunk 4 above)
Comment (loadtheaccumulator) [reply pkg/clients/pulp/guards_turnpike.go:90]: Yeah. Actually, I think I might be able to just use HeaderGuard in place of TurnpikeGuard.

Comment (loadtheaccumulator) [reply pkg/clients/pulp/guards_turnpike.go:90]: I'll look closer in the AM, but just some minor changes to HeaderGuard since Turnpike is not based just on orgid will probably serve both.

File: pkg/clients/pulp/guards_composite.go [hunk 6]
```diff
@@ -65,12 +66,44 @@ func (ps *PulpService) CompositeGuardReadByOrgID(ctx context.Context, orgID stri
...
+	}
+
+	logrus.Debug("Content Guard has the expected hrefs")
+
+	return true
+}
+
 // CompositeGuardEnsure ensures that the composite guard is created and returns it. The method is idempotent.
-func (ps *PulpService) CompositeGuardEnsure(ctx context.Context, orgID, headerHref, rbacHref string) (*CompositeContentGuardResponse, error) {
+func (ps *PulpService) CompositeGuardEnsure(ctx context.Context, orgID string, pulpGuardHrefs []string) (*CompositeContentGuardResponse, error) {
```
Comment (lzap) [inline pkg/clients/pulp/guards_composite.go:101]: Nitpick: This could be variadic function.

File: pkg/clients/pulp/guards_composite.go [hunk 7]
```diff
@@ -65,12 +66,44 @@ func (ps *PulpService) CompositeGuardReadByOrgID(ctx context.Context, orgID stri
 	return ps.CompositeGuardRead(ctx, id)
 }
 
+func contentGuardHrefsAreEqual(cghrefa, cghrefb []string) bool {
```
Comment (lzap) [inline pkg/clients/pulp/guards_composite.go:69]: I suggest a unit test for this one.

File: pkg/clients/pulp/guards_composite.go [hunk 8]
```diff
@@ -65,12 +66,44 @@ func (ps *PulpService) CompositeGuardReadByOrgID(ctx context.Context, orgID stri
...
+	// compare content guard hrefs (actual vs. expected)
+	for i := range cghrefa {
+		if !slices.Contains(cghrefb, cghrefa[i]) {
+			logrus.WithField("href_mismatch", cghrefa[i]).Warning("Content Guard href mismatch")
+
+			return false
+		}
+	}
+
+	logrus.Debug("Content Guard has the expected hrefs")
```
Comment (lzap) [inline pkg/clients/pulp/guards_composite.go:95]: Generally, it is often better to avoid positive (even debug) statements that carry zero information.

File: pkg/clients/pulp/guards_composite.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (lzap) [reply pkg/clients/pulp/guards_composite.go:165]: No, feel free to get rid of RBAC right here in this patch. This was a dead end which I was exploring and turned out not to be viable. Sorry!

File: pkg/clients/pulp/guards_turnpike.go (hunk 4 above)
Comment (lzap) [reply pkg/clients/pulp/guards_turnpike.go:90]: If there was a chance to generalize the HeaderGuard instead that would be ideal. I am fine with this.

Tho I would suggest to move the header value up one call stack above, feels more right. It would help with testability, but there are no tests other than integration tests for this package (which I thought woudl be enough). So fine I guess.

File: pkg/clients/pulp/guards_composite.go (hunk 7 above)
Comment (loadtheaccumulator) [reply pkg/clients/pulp/guards_composite.go:69]: Already in the works. :-)

File: pkg/clients/pulp/guards_composite.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (loadtheaccumulator) [reply pkg/clients/pulp/guards_composite.go:165]: Cool. I'll remove it when cert is working upstream, so I can keep moving.

File: pkg/clients/pulp/guards_composite.go (hunk 8 above)
Comment (loadtheaccumulator) [reply pkg/clients/pulp/guards_composite.go:95]: I'd like to return an error and have the calling func take care of logging. Right now looking to see everything work.

File: pkg/clients/pulp/guards_composite.go (hunk 6 above)
Comment (loadtheaccumulator) [reply pkg/clients/pulp/guards_composite.go:101]: I'm thinking since nothing really does anything with the hrefs other than pass them on, that variadic isn't necessary in the stack and it could all just be a slice.

File: pkg/clients/pulp/guards_composite.go (hunk 8 above)
Comment (lzap) [reply pkg/clients/pulp/guards_composite.go:95]: Correction, this one is okay, the one above is little bit less informative. But since this is all `DEBUG` I am good.

File: pkg/clients/pulp/guards_composite.go (hunk 6 above)
Comment (loadtheaccumulator) [reply pkg/clients/pulp/guards_composite.go:101]: But either way (or both) work for me. We can nail it down while in draft.

File: pkg/clients/pulp/guards_turnpike.go [hunk 9]
```diff
@@ -0,0 +1,172 @@
+package pulp
```
Comment (ezr-ondrej) [inline pkg/clients/pulp/guards_turnpike.go:1]: Could you add simple explanation of Turnpike, or link to explanation as a header comment in this file?

Nothing big, few words :)

**Comment by lzap** [comment]
My bad, is taking too long. I am almost there, the patch is in, we need to kill workers...

**Comment by lzap** [comment]
The update edge builds are finally fixed on stage. Let me know when this is ready for re-review.

File: pkg/clients/pulp/guards_composite.go [hunk 10]
```diff
@@ -107,12 +108,27 @@ func (ps *PulpService) ContentGuardEnsure(ctx context.Context, orgID string) (*C
 		return nil, err
 	}
 
-	rcg, err := ps.RbacGuardEnsure(ctx)
-	if err != nil {
-		return nil, err
+	var contentGuardHref *string
+
+	if feature.PulpIntegrationTurnpikeGuard.IsEnabled() {
```
[the remark this answers is not in the document]
Comment (loadtheaccumulator) [reply pkg/clients/pulp/guards_composite.go]: It's set as empty string in defaults, but can add for good measure

[the remark this answers is not in the document]
Comment (loadtheaccumulator) [reply pkg/clients/pulp/guards_composite.go]: I added the check and error in TurnpikeGuardCreate()
